diff -purN a/dlls/appwiz.cpl/addons.c b/dlls/appwiz.cpl/addons.c
--- a/dlls/appwiz.cpl/addons.c	2021-11-25 10:00:17.793874353 +0200
+++ b/dlls/appwiz.cpl/addons.c	2021-11-25 10:19:13.600502190 +0200
@@ -57,10 +57,10 @@ WINE_DEFAULT_DEBUG_CHANNEL(appwizcpl);
 #define GECKO_SHA "???"
 #endif
 
-#define MONO_VERSION "6.4.1"
+#define MONO_VERSION "7.0.0"
 #if defined(__i386__) || defined(__x86_64__)
 #define MONO_ARCH "x86"
-#define MONO_SHA "e7cb533e0ca6ff2753c1449661be5c8833522f156fd36b48fa5d881cf0354270"
+#define MONO_SHA "b37e6fc9e590e582243dc25d72a5fcc330c3a7970dfdc98a7a81d23845ba8900"
 #else
 #define MONO_ARCH ""
 #define MONO_SHA "???"
@@ -96,7 +96,7 @@ static const addon_info_t addons_info[]
         L"wine-mono-" MONO_VERSION "-" MONO_ARCH ".msi",
         L"mono",
         MONO_SHA,
-        "https://github.com/madewokherd/wine-mono/releases/download/wine-mono-" MONO_VERSION" /wine-mono-" MONO_VERSION "-" MONO_ARCH ".msi",
+        "https://github.com/madewokherd/wine-mono/releases/download/wine-mono-" MONO_VERSION "/wine-mono-" MONO_VERSION "-" MONO_ARCH ".msi",
         "Dotnet", "MonoUrl", "MonoCabDir",
         MAKEINTRESOURCEW(ID_DWL_MONO_DIALOG)
     }
diff -purN a/dlls/mscoree/mscoree_private.h b/dlls/mscoree/mscoree_private.h
--- a/dlls/mscoree/mscoree_private.h	2021-11-25 10:34:29.723804202 +0200
+++ b/dlls/mscoree/mscoree_private.h	2021-11-25 10:57:55.973756128 +0200
@@ -45,7 +45,7 @@ extern HRESULT assembly_get_runtime_vers
 extern HRESULT assembly_get_vtable_fixups(ASSEMBLY *assembly, VTableFixup **fixups, DWORD *count) DECLSPEC_HIDDEN;
 extern HRESULT assembly_get_native_entrypoint(ASSEMBLY *assembly, NativeEntryPointFunc *func) DECLSPEC_HIDDEN;
 
-#define WINE_MONO_VERSION "6.4.1"
+#define WINE_MONO_VERSION "7.0.0"
 
 /* Mono embedding */
 typedef struct _MonoDomain MonoDomain;
