diff -purN a/Makefile b/Makefile
--- a/Makefile	2021-07-27 13:15:15.508205146 +0300
+++ b/Makefile	2021-08-18 12:27:33.927144096 +0300
@@ -1820,7 +1820,7 @@ install-manpages:: manpages
 # Rules for generated source files
 
 dlls/ntdll/unix/version.c: dummy
-	@version=`(GIT_DIR=$(srcdir)/.git git describe HEAD 2>/dev/null || echo "wine-$(PACKAGE_VERSION)") | sed -n -e '$$s/\(.*\)/const char wine_build[] = "\1";/p'` && (echo $$version | cmp -s - $@) || echo $$version >$@ || ($(RM) $@ && exit 1)
+	@version=`echo "wine-$(PACKAGE_VERSION)" | sed -n -e '$$s/\(.*\)/const char wine_build[] = "\1 (Proton)";/p'` && (echo $$version | cmp -s - $@) || echo $$version >$@ || ($(RM) $@ && exit 1)
 
 programs/winetest/build.rc: dummy
 	@build="STRINGTABLE { 1 \"`GIT_DIR=$(srcdir)/.git git rev-parse HEAD 2>/dev/null`\" }" && (echo $$build | cmp -s - $@) || echo $$build >$@ || (rm -f $@ && exit 1)
