diff -purN a/dlls/ucrtbase/Makefile.in b/dlls/ucrtbase/Makefile.in
--- a/dlls/ucrtbase/Makefile.in	2021-07-27 14:02:11.350697988 +0300
+++ b/dlls/ucrtbase/Makefile.in	2021-07-27 14:32:30.105785474 +0300
@@ -37,6 +37,7 @@ C_SRCS = \
 	printf.c \
 	process.c \
 	scanf.c \
+	sincos.c \
 	string.c \
 	thread.c \
 	time.c \
diff -purN a/dlls/ucrtbase/sincos.c b/dlls/ucrtbase/sincos.c
--- a/dlls/ucrtbase/sincos.c	1970-01-01 03:00:00.000000000 +0300
+++ b/dlls/ucrtbase/sincos.c	2021-07-27 15:19:14.597511610 +0300
@@ -0,0 +1,19 @@
+
+/* this function is part of the import lib to provide floating */
+#if 0
+#pragma makedep implib
+#endif
+
+#include <math.h>
+
+#pragma GCC optimize ("O0")
+
+void __cdecl sincos(double a, double* s, double* c) {
+  *s = sin(a);
+  *c = cos(a);
+}
+
+void __cdecl sincosf(float a, float* s, float* c) {
+  *s = sinf(a);
+  *c = cosf(a);
+}
