diff -purN a/tools/widl/widl.c b/tools/widl/widl.c
--- a/tools/widl/widl.c	2022-04-21 23:46:36.874356928 +0300
+++ b/tools/widl/widl.c	2022-04-22 00:03:41.032369548 +0300
@@ -174,7 +174,7 @@ enum {
 };
 
 static const char short_options[] =
-    "b:cC:d:D:EhH:I:L:m:No:O:pP:rsS:tT:uU:VW";
+    "b:cC:d:D:EhH:I:L:m:No:O:p::P:rsS:tT:uU:VW";
 static const struct long_option long_options[] = {
     { "acf", 1, ACF_OPTION },
     { "app_config", 0, APP_CONFIG_OPTION },
@@ -624,6 +624,7 @@ static void option_callback( int optc, c
       else error( "Invalid argument '-O%s'\n", optarg );
       break;
     case 'p':
+      if (optarg && strcmp(optarg, "thread") == 0) break;
       do_everything = 0;
       do_proxies = 1;
       break;
