diff -purN a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
--- a/dlls/ntdll/signal_x86_64.c	2019-11-06 18:04:35.000000000 +0200
+++ b/dlls/ntdll/signal_x86_64.c	2019-11-16 17:51:44.987670561 +0200
@@ -2645,6 +2645,8 @@ static NTSTATUS call_stack_handlers( EXC
     UNWIND_HISTORY_TABLE table;
     DISPATCHER_CONTEXT dispatch;
     CONTEXT context;
+    MEMORY_BASIC_INFORMATION wine_frame_stack_info, current_stack_info;
+    int is_teb_frame_in_current_stack = 1;
     LDR_MODULE *module;
     NTSTATUS status;
 
@@ -2652,6 +2654,13 @@ static NTSTATUS call_stack_handlers( EXC
     dispatch.TargetIp      = 0;
     dispatch.ContextRecord = &context;
     dispatch.HistoryTable  = &table;
+
+    if ( !(NtQueryVirtualMemory(NtCurrentProcess(), teb_frame, MemoryBasicInformation, &wine_frame_stack_info, sizeof(MEMORY_BASIC_INFORMATION), NULL)) &&
+         !(NtQueryVirtualMemory(NtCurrentProcess(), (PVOID)context.Rsp, MemoryBasicInformation, &current_stack_info, sizeof(MEMORY_BASIC_INFORMATION), NULL)))
+    {
+        is_teb_frame_in_current_stack = wine_frame_stack_info.AllocationBase == current_stack_info.AllocationBase;
+    }
+
     for (;;)
     {
         /* FIXME: should use the history table to make things faster */
@@ -2753,7 +2762,7 @@ static NTSTATUS call_stack_handlers( EXC
             }
         }
         /* hack: call wine handlers registered in the tib list */
-        else while ((ULONG64)teb_frame < context.Rsp)
+        else if (is_teb_frame_in_current_stack) while ((ULONG64)teb_frame < context.Rsp)
         {
             TRACE( "found wine frame %p rsp %lx handler %p\n",
                     teb_frame, context.Rsp, teb_frame->Handler );
